# Sequence analysis of Kappa and Delta variants in Hong Kong

*Haogao Gu, Samuel S.M. Cheng, Pavithra Krishnan, Daisy Y.M. Ng, Lydia D.J Chang, Gigi Y.Z. Liu, Sammi S.Y. Cheuk, Mani M.Y. Hui, Mathew C.Y. Fan, Jacob H.L. Wan, Leo H.K. Lau, Daniel K.W. Chu, Vijaykrishna Dhanasekaran, Malik Peiris, and Leo L.M. Poon*

---

[The paper is now available on the EID journal.](https://doi.org/10.3201/eid2801.211804)

## Abstract
We sequenced 50% of COVID-19 cases recently imported to Hong Kong (N=212) and identified 70 cases infected by the Delta variants. The phylogeny of our Delta variant sequences is similar to the one built by using the global dataset, suggesting travel hubs can be used to monitor SARS-CoV-2 sequence diversity. 

## Description
The source codes for different analyses were stored in different sub-directories under the scripts directory. The results generated by the scripts can be found in the results directory.

## Environment
```r
r$> sessionInfo()          
R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 11.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] C/UTF-8/C/C/C/C

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] gdtools_0.2.3       officer_0.3.19      ggnewscale_0.4.5    colorspace_2.0-2    ggrepel_0.9.1      
 [6] ggtree_3.0.2        lubridate_1.7.10    Biostrings_2.60.1   GenomeInfoDb_1.28.1 XVector_0.32.0     
[11] IRanges_2.26.0      S4Vectors_0.30.0    BiocGenerics_0.38.0 forcats_0.5.1       stringr_1.4.0      
[16] dplyr_1.0.7         purrr_0.3.4         readr_1.4.0         tidyr_1.1.3         tibble_3.1.2       
[21] ggplot2_3.3.5       tidyverse_1.3.1    

loaded via a namespace (and not attached):
 [1] nlme_3.1-152           bitops_1.0-7           fs_1.5.0               httr_1.4.2            
 [5] rprojroot_2.0.2        tools_4.1.0            backports_1.2.1        utf8_1.2.1            
 [9] R6_2.5.0               DBI_1.1.1              lazyeval_0.2.2         withr_2.4.2           
[13] tidyselect_1.1.1       phangorn_2.7.1         compiler_4.1.0         cli_3.0.0             
[17] rvest_1.0.0            xml2_1.3.2             labeling_0.4.2         scales_1.1.1          
[21] quadprog_1.5-8         systemfonts_1.0.2      httpgd_1.1.1           digest_0.6.27         
[25] pkgconfig_2.0.3        dbplyr_2.1.1           rvg_0.2.5              rlang_0.4.11          
[29] readxl_1.3.1           rstudioapi_0.13        farver_2.1.0           generics_0.1.0        
[33] jsonlite_1.7.2         zip_2.2.0              RCurl_1.98-1.3         magrittr_2.0.1        
[37] GenomeInfoDbData_1.2.6 patchwork_1.1.1        Matrix_1.3-4           Rcpp_1.0.7            
[41] munsell_0.5.0          fansi_0.5.0            ape_5.5                lifecycle_1.0.0       
[45] stringi_1.7.2          zlibbioc_1.38.0        grid_4.1.0             crayon_1.4.1          
[49] lattice_0.20-44        haven_2.4.1            hms_1.1.0              pillar_1.6.1          
[53] uuid_0.1-4             igraph_1.2.6           codetools_0.2-18       fastmatch_1.1-0       
[57] reprex_2.0.0           glue_1.4.2             BiocManager_1.30.16    modelr_0.1.8          
[61] vctrs_0.3.8            treeio_1.16.1          cellranger_1.1.0       gtable_0.3.0          
[65] assertthat_0.2.1       broom_0.7.8            tidytree_0.3.4         later_1.2.0           
[69] aplot_0.0.6            rvcheck_0.1.8          ellipsis_0.3.2         here_1.0.1           
```